docs/docs/state-and-lifecycle.md

```diff
@@ -0,0 +1,504 @@
+    this.state = {date: new Date()};
+  }
+
+  componentDidMount() {
+    this.intervalId = setInterval(this.tick, 1000);
+  }
+
+  componentWillUnmount() {
+    clearInterval(this.intervalId);
+  }
+
+  tick() {
+    this.setState({
+      date: new Date()
+    });
+  }
+
+  render() {
+    return (
+      <div>
+        <h1>Hello, world!</h1>
+        <h2>It is {this.state.date.toLocaleTimeString()}.</h2>
+      </div>
+    );
+  }
+}
+
+const container = document.getElementById('root');
+ReactDOM.render(<Clock />, container);
+```
+
+[Try it on Codepen.](http://codepen.io/gaearon/pen/amqdNA?editors=0010)
+
+Now the clock ticks every second.
+
+## A Note on Binding Methods
+
+Note how we needed to bind `this.tick` in the constructor:
+
+```js{3}
+  constructor(props) {
+    super(props);
+    this.tick = this.tick.bind(this);
+    this.state = {date: new Date()};
+  }
+```
+
+In JavaScript, class methods are not [bound](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_objects/Function/bind) by default. If you forget to bind `this.tick` and pass it to `setInterval()`, `this` inside it will be `undefined`.
+
+This is not React-specific behavior, and is a part of [how functions work in JavaScript](https://www.smashingmagazine.com/2014/01/understanding-javascript-function-prototype-bind/).
+
+Generally, the rule of thumb is that if you refer to a method without `()` after it, such as `setTimeout(this.tick, 1000)`, you need to bind that method.
+
+You don't need to bind `render()` or any of the lifecycle methods.
+
+## Using the State Correctly
+
+There are three things you should know about `setState()`.
+
+### Do Not Modify State Directly
+
+For example, this will not work:
+
+```js
+// Wrong
+this.state.comment = 'Hello';
+```
+
+Instead, use `setState()`:
+
+```js
+// Correct
+this.setState({comment: 'Hello'});
+```
+
+### setState() Is Asynchronous
+
+React may batch multiple `setState()` calls into a single update for performance.
+
+Because `this.props` and `this.state` always represent the currently rendered state and props, you should not rely on their values when updating the state.
+
+For example, this code may fail to update the counter:
+
+```js
+// Wrong
+this.setState({
+  counter: this.state.counter + this.props.increment,
+});
+```
+
+To fix it, use a second form of `setState()` that accepts a function rather than an object. That function will receive the previous state as the first argument, and the current props as the second argument:
+
+```js
+// Correct
+this.setState((prevState, props) => ({
+  counter: prevState.counter + props.increment
+}));
+```
+
+We used an [arrow function](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Functions/Arrow_functions) above, but it also works with regular functions:
+
+```js
+// Correct
+this.setState(function(prevState, props) {
+  return {
+    counter: prevState.counter + props.increment
+  };
+});
+```
+
+### setState() Merges the State Shallowly
```

Comment:
I don't think "shallow merging" is a commonly used phrase. How about describing more precisely what shallow merging means. Something like, if you only provide some keys in setState, then setState only updates those keys. So if state is {x: 1} and you call setState({y: 2}) your new state is {x: 1, y: 2}.